File: report/src/main/java/com/vmlens/report/runelementtype/memoryaccesskey/AtomicMethodIdAndObjectHashcode.java
package com.vmlens.report.runelementtype.memoryaccesskey;

import com.vmlens.report.description.DescriptionContext;
import com.vmlens.report.description.NeedsDescriptionCallback;

public class AtomicMethodIdAndObjectHashcode implements MemoryAccessKey  {

    private final int atomicMethodId;
    private final long objectHashCode;

    public AtomicMethodIdAndObjectHashcode(int atomicMethodId, long objectHashCode) {
        this.atomicMethodId = atomicMethodId;
        this.objectHashCode = objectHashCode;
    }

    @Override
    public String asString(DescriptionContext context) {
        return context.methodName(atomicMethodId) + "@" + objectHashCode;
    }

    @Override
    public void addToNeedsDescription(NeedsDescriptionCallback callback) {
        callback.needsMethod(atomicMethodId);
    }

}

File: sync-bug/src/main/scala/com/anarsoft/race/detection/report/element/runelementtype/memoryaccesskey/AtomicMethodIdAndObjectHashcode.scala
package com.anarsoft.race.detection.report.element.runelementtype.memoryaccesskey

import com.anarsoft.race.detection.report.description.{DescriptionContext, NeedsDescriptionCallback}


class AtomicMethodIdAndObjectHashcode(val atomicMethodId: Int, val objectHashCode: Long)  
  extends GenericMemoryAccessKey[AtomicMethodIdAndObjectHashcode] {
  override def asString(context: DescriptionContext): String = context.methodNameWithoutSource(atomicMethodId)

  override def addToNeedsDescription(callback: NeedsDescriptionCallback): Unit = {
    callback.needsMethod(atomicMethodId)
  }

  override def objectHashCodeOption: Option[Long] = Some(objectHashCode);


  private def canEqual(other: Any): Boolean = other.isInstanceOf[AtomicMethodIdAndObjectHashcode]
  
  override def equals(other: Any): Boolean = other match {
    case that: AtomicMethodIdAndObjectHashcode =>
      that.canEqual(this) &&
        atomicMethodId == that.atomicMethodId &&
        objectHashCode == that.objectHashCode
    case _ => false
  }
  
  override def hashCode(): Int = {
    val state = Seq(atomicMethodId, objectHashCode)
    state.map(_.hashCode()).foldLeft(0)((a, b) => 31 * a + b)
  }

  override def compare(that: AtomicMethodIdAndObjectHashcode): Int = {
    if(objectHashCode != that.objectHashCode) {
      objectHashCode.compare(that.objectHashCode)
    } else {
      atomicMethodId.compare(that.atomicMethodId)
    }
  }
}
